### PR TITLE
FEATURE: BCO-1119 Introduce data sources API to GDC Ruby SDK

### PR DESCRIPTION
**Pull request by tuannnh**
@Seikitsu Could you approve this PR?

### CODE REVIEW

**Comment by danh-ung** [comment]
PR, commit message MUST link to a JIRA ticket.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/514/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/517/ : FAILURE in 7h 13m 47s

**Comment by tuannnh** [comment]
@danh-ung Hi, this PR linked to this JIRA ticket: [https://jira.intgdc.com/browse/BCO-1119](url)

**Comment by danh-ung** [comment]
@tuannnh PR and commit message have to follow commit policy, see https://confluence.intgdc.com/display/pap/Commit+Policy+Examples. In short, your PR and commit message should be "FEATURE: BCO-1119 Introduce data sources API to GDC Ruby SDK"

**Comment by tuannnh** [comment]
Hi @danh-ung, thank you for supporting. I have changed the commit message and PR. Is it anything else I have to follow?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/515/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/518/ : SUCCESS in 21m 06s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/516/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/519/ : SUCCESS in 20m 51s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/517/ : SUCCESS in 5s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/520/ : SUCCESS in 21m 06s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/518/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/521/ : SUCCESS in 20m 21s

File: lib/gooddata/models/data_source.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+#
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
```
Comment (hung-nguyen-hoang) [inline lib/gooddata/models/data_source.rb:4]: Please update the copyright to "# Copyright (c) 2010-2020 GoodData Corporation. All rights reserved."

File: lib/gooddata/models/data_source.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        uri = '/gdc/dataload/dataSources'
+        request_value = {}
+        request_value[:dataSource] = data_source
+        client.post(uri, request_value)
+      end
+
+      # @param [Hash] data_source The data source json data
+      # @param [Object] client The Rest Client object
+      def redeploy_data_source(data_source, client)
+        data_source_id = data_source[:id]
```
Comment (hung-nguyen-hoang) [inline lib/gooddata/models/data_source.rb:50]: Please verify the id existing in data source json data

**Comment by hung-nguyen-hoang** [comment]
Currently the code already existing the data source APIs in https://github.com/gooddata/gooddata-ruby/blob/master/lib/gooddata/helpers/data_source_helpers.rb So please help refactor move the data source APIs existing in data_source_helpers.rb to the new data sources API.
Thank you.

**Comment by tuannnh** [comment]
> Currently the code already existing the data source APIs in https://github.com/gooddata/gooddata-ruby/blob/master/lib/gooddata/helpers/data_source_helpers.rb So please help refactor move the data source APIs existing in data_source_helpers.rb to the new data sources API.
> Thank you.

After moving the APIs from data_source_helpers.rb file. Can I remove that file?

**Comment by hung-nguyen-hoang** [comment]
Yes, please remove it. Thank you.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/519/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/522/ : FAILURE in 6m 18s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/520/ : SUCCESS in 4s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/523/ : SUCCESS in 26m 22s

File: lib/gooddata/models/data_source.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+#
+# Copyright (c) 2010-2020 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+module GoodData
+  class DataSource < Rest::Resource
+    class << self
```
Comment (danh-ung) [inline lib/gooddata/models/data_source.rb:10]: models/data_source.rb should represent for dataSource object with functions to CRUD:
- constructor: create new datasource object or load it from server.
- create: create new datasource.
- save: create or update datasource
- all: list data sources

You can inspire from other resources process.rb or schedule.rb

**Review by danh-ung — CHANGES_REQUESTED** [review]

**Comment by danh-ung** [comment]
Please add automation tests, see similar spec/integration/models/schedule_spec.rb

**Comment by danh-ung** [comment]
Closed, replaced by https://github.com/gooddata/gooddata-ruby/pull/1714

**Pull request closed without merging**